src/controller/ExampleOperationalCredentialsIssuer.cpp

```diff
@@ -411,6 +411,88 @@ CHIP_ERROR ExampleOperationalCredentialsIssuer::GenerateNOCChain(const ByteSpan
+{
+    uint8_t rcac[Credentials::kMaxDERCertLength];
+    MutableByteSpan rcacSpan = MutableByteSpan(rcac);
+    ChipDN rcac_dn;
+    CHIP_ERROR err = CHIP_NO_ERROR;
+
+    P256PublicKey pubkey;
+    ReturnErrorOnFailure(VerifyCertificateSigningRequest(icaCsr.data(), icaCsr.size(), pubkey));
+
+    uint16_t rcacBufLen = static_cast<uint16_t>(std::min(rcacSpan.size(), static_cast<size_t>(UINT16_MAX)));
+    PERSISTENT_KEY_OP(mIndex, kOperationalCredentialsRootCertificateStorage, key,
+                      err = mStorage->SyncGetKeyValue(key, rcacSpan.data(), rcacBufLen));
+    ReturnErrorOnFailure(err);
+
+    {
+        uint64_t rcacId;
+        // Found root certificate in the storage.
+        rcacSpan.reduce_size(rcacBufLen);
+        ReturnErrorOnFailure(ExtractSubjectDNFromX509Cert(rcacSpan, rcac_dn));
+        ReturnErrorOnFailure(rcac_dn.GetCertChipId(rcacId));
+        VerifyOrReturnError(rcacId == mRootIssuerId, CHIP_ERROR_INTERNAL);
+    }
+
+    ChipDN icac_dn;
+    ReturnErrorOnFailure(icac_dn.AddAttribute_MatterICACId(mIntermediateIssuerId));
+
+    ChipLogProgress(Controller, "Signing ICA CSR");
+
+    ReturnErrorOnFailure(
+        IssueX509Cert(mNow, mValidity, rcac_dn, icac_dn, CertType::kIcac, mUseMaximallySizedCerts, pubkey, mRootIssuer, icac));
+
+    VerifyOrReturnError(CanCastTo<uint16_t>(icac.size()), CHIP_ERROR_INTERNAL);
+
+    PERSISTENT_KEY_OP(mIndex, kOperationalCredentialsIntermediateCertificateStorage, key,
+                      ReturnErrorOnFailure(mStorage->SyncSetKeyValue(key, icac.data(), static_cast<uint16_t>(icac.size()))));
+
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR ExampleOperationalCredentialsIssuer::SignICAC(const ByteSpan & icaCsr, Callback::Callback<OnICACSigned> * onCompletion)
+{
+    VerifyOrReturnError(mInitialized, CHIP_ERROR_UNINITIALIZED);
+
+    chip::Platform::ScopedMemoryBuffer<uint8_t> icac;
+    VerifyOrReturnError(icac.Alloc(kMaxDERCertLength), CHIP_ERROR_NO_MEMORY);
+    MutableByteSpan icacSpan(icac.Get(), kMaxDERCertLength);
+
+    ReturnErrorOnFailure(SignICACAfterValidation(icaCsr, icacSpan));
+
+    // Callback onto commissioner.
+    ChipLogProgress(Controller, "Providing signed ICA to the commissioner");
+    onCompletion->mCall(onCompletion->mContext, CHIP_NO_ERROR, icacSpan);
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR ExampleOperationalCredentialsIssuer::SignNOC(const ByteSpan & icac, const ByteSpan & nocCsr, MutableByteSpan & noc)
+{
+    P256PublicKey pubkey;
+    ReturnErrorOnFailure(VerifyCertificateSigningRequest(nocCsr.data(), nocCsr.size(), pubkey));
+
+    ChipDN icac_dn = ChipDN{};
+    ReturnErrorOnFailure(ExtractSubjectDNFromX509Cert(icac, icac_dn));
+
+    PERSISTENT_KEY_OP(mIndex, kOperationalCredentialsIntermediateCertificateStorage, key,
+                      ReturnErrorOnFailure(mStorage->SyncSetKeyValue(key, icac.data(), static_cast<uint16_t>(icac.size()))));
```

Comment:
Wait, how does this work?  What if the provided icac does not match mIntermediateIssuer?